src/dynamicprompts/samplers/combinatorial.py

```diff
@@ -32,136 +30,113 @@ def _dedupe(arr: list[str]) -> tuple[str, ...]:
 
 
 def _combo_to_prompt(
-    sampler_router: SamplerRouter,
+    sampling_context: SamplingContext,
     combo: list[Command],
 ) -> typing.Iterable[list[str]]:
     if len(combo) == 0:
         yield []
         return
 
     c_1, c_rest = combo[0], combo[1:]
-    gen = sampler_router.generator_from_command(c_1)
-    if c_1.sampling_method != SamplingMethod.COMBINATORIAL:
-        for rest_prompt in _combo_to_prompt(sampler_router, c_rest):
+    gen = sampling_context.generator_from_command(c_1)
+    if sampling_context.get_effective_sampling_method(c_1).is_nonfinite():
+        for rest_prompt in _combo_to_prompt(sampling_context, c_rest):
             val = next(gen)
             yield [val] + rest_prompt
     else:
         for p in gen:
-            for rest_prompt in _combo_to_prompt(sampler_router, c_rest):
+            for rest_prompt in _combo_to_prompt(sampling_context, c_rest):
                 if rest_prompt:
                     yield [p] + rest_prompt
                 else:
                     yield [p]
 
 
 class CombinatorialSampler(Sampler):
-    def __init__(
+    def _get_sequence(
         self,
-        *,
-        wildcard_manager: WildcardManager,
-        ignore_whitespace: bool = False,
-        sampler_router: SamplerRouter,
-        parser_config: ParserConfig = default_parser_config,
-    ):
-        super().__init__(
-            wildcard_manager=wildcard_manager,
-            ignore_whitespace=ignore_whitespace,
-            sampler_router=sampler_router,
-            parser_config=parser_config,
-        )
-        self._sampler_router = sampler_router
-
-    def _propagate_sampling_method(self, commands: Iterable[Command]) -> None:
-        for cmd in commands:
-            if cmd.sampling_method == SamplingMethod.DEFAULT:
-                cmd.sampling_method = SamplingMethod.COMBINATORIAL
-
-    def _get_sequence(self, command: SequenceCommand) -> StringGen:
-        self._propagate_sampling_method(command.tokens)
-
-        sentinel_start = LiteralCommand(
-            "",
-            sampling_method=SamplingMethod.COMBINATORIAL,
-        )
-        sentinel_end = LiteralCommand("", sampling_method=SamplingMethod.COMBINATORIAL)
-
+        command: SequenceCommand,
+        context: SamplingContext,
+    ) -> StringGen:
         non_combo_commands = [
             c
             for c in command.tokens
             if c.sampling_method != SamplingMethod.COMBINATORIAL
         ]
         command_collection = CommandCollection(
             non_combo_commands,
-            self._sampler_router,
+            context=context,
```

Comment:
`non_combo_commands` overloads the word combo - here it means combinatorial rather than combination. Perhaps we can renamed it to `non_combinatorial_commands`. Or perhaps it would be better as `non_finite_commands`.